src/sentry/integrations/gitlab/webhooks.py

```diff
@@ -79,92 +136,68 @@ def create_commits(self, event, organization, repo):
-
         return super(GitlabWebhookEndpoint, self).dispatch(request, *args, **kwargs)
 
-    def post(self, request, organization_id):
+    def post(self, request):
         try:
-            organization = Organization.objects.get_from_cache(
-                id=organization_id,
-            )
-        except Organization.DoesNotExist:
-            logger.error(
-                PROVIDER_NAME + '.webhook.invalid-organization',
+            integration = Integration.objects.filter(
+                provider=self.provider,
+                external_id=request.META['HTTP_X_GITLAB_TOKEN']
+            ).prefetch_related('organizations').get()
+        except Integration.DoesNotExist:
+            logger.info(
+                'gitlab.webhook.invalid-organization',
                 extra={
-                    'organization_id': organization_id,
+                    'external_id': request.META['HTTP_X_GITLAB_TOKEN'],
                 }
             )
             return HttpResponse(status=400)
 
-        body = six.binary_type(request.body)
-        if not body:
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-body', extra={
-                    'organization_id': organization.id,
+        if integration.organizations.count() != 1:
+            logger.info(
+                'gitlab.webhook.extra-organizations',
+                extra={
+                    'count': len(integration.organizations),
+                    'external_id': integration.external_id,
                 }
             )
             return HttpResponse(status=400)
 
         try:
-            handler = self.get_handler(request.META['HTTP_X_EVENT_KEY'])
-        except KeyError:
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-event', extra={
-                    'organization_id': organization.id,
+            event = json.loads(request.body.decode('utf-8'))
+        except JSONDecodeError:
+            logger.info(
+                'gitlab.webhook.invalid-json',
+                extra={
+                    'external_id': integration.external_id
                 }
             )
             return HttpResponse(status=400)
 
-        if not handler:
-            return HttpResponse(status=204)
-
         try:
-            event = json.loads(body.decode('utf-8'))
-        except JSONDecodeError:
-            logger.error(
-                PROVIDER_NAME + '.webhook.invalid-json',
-                extra={
-                    'organization_id': organization.id,
-                },
-                exc_info=True
-            )
+            handler = self._handlers[request.META['HTTP_X_GITLAB_EVENT']]
+        except KeyError:
+            logger.info('gitlab.webhook.missing-event', extra={
+                'event': request.META['HTTP_X_GITLAB_EVENT']
+            })
             return HttpResponse(status=400)
 
-        handler()(organization, event)
+        handler()(integration, integration.organizations.first(), event)
```

Comment:
So, when you use `first` that is the equivalent of querying and then using an order by. I can't actually recall how organizations are ordered, but I have a feeling the default behavior isn't probably what you would want. 

https://django.readthedocs.io/en/1.6.x/ref/models/querysets.html?#first 

Is there a reason the organizations are being sorted, or do you simply just want the first element?

Reply:
I'm using `.first()` as `integration.organizations` is a Many To Many association and I couldn't use `[0]` on the property because `ManyToManyManager` doesn't support that interface. I'm not interested in the order, I just want one record from the relation.